File: vscode-extension/src/test/services/StateHelper.test.ts
import * as assert from 'assert';
import * as path from 'path';
import { StateHelper } from '../../services/StateHelper';
import { ActiveFile, State } from '../../types';
import { IdeMetaData } from '../../IdeMetaData';

/**
 * Unit tests for StateHelper utility.
 */
describe('StateHelper', () => {
    describe('buildState', () => {
        it('should build a flat state with normalized paths', () => {
            const rootPath = '/User/Project';
            const openedFiles: string[] = [
                '/User/Project/src/main.ts',
                '/User/Project/README.md'
            ];
            const activeFile: ActiveFile = {
                filePath: '/User/Project/src/main.ts',
                cursor: 10,
                column: 5
            };

            const state = StateHelper.buildState(rootPath, openedFiles, activeFile);

            const rootPathLower = path.resolve(rootPath).toLowerCase();
            assert.strictEqual(state.root, rootPathLower);
            assert.strictEqual(state.editorState.openedFiles.length, 2);

            const file1 = state.editorState.openedFiles.find(f => f.endsWith('main.ts'));
            assert.ok(file1);
            
            const file2 = state.editorState.openedFiles.find(f => f.endsWith('readme.md'));
            assert.ok(file2);

            const active = state.editorState.activeFile;
            assert.ok(active);
            assert.strictEqual(active?.filePath, path.resolve('/User/Project/src/main.ts').toLowerCase());
            assert.strictEqual(active?.cursor, 10);
            assert.strictEqual(active?.column, 5);
        });

        it('should filter out files outside of root', () => {
            const rootPath = '/User/Project';
            const openedFiles: string[] = [
                '/User/Project/in.ts',
                '/User/Other/out.ts'
            ];

            const state = StateHelper.buildState(rootPath, openedFiles, undefined);
            assert.strictEqual(state.editorState.openedFiles.length, 1);
            assert.ok(state.editorState.openedFiles[0].includes('in.ts'));
        });
    });

    describe('getFiles', () => {
        it('should return files belonging to root', () => {
            const rootPath = '/User/Project';
            const state: State = {
                timestamp: Date.now(),
                source: 'test',
                ide: 'vscode',
                root: path.resolve(rootPath).toLowerCase(),
                editorState: {
                    openedFiles: [
                        path.resolve('/User/Project/file1.ts').toLowerCase(),
                        path.resolve('/User/Other/file2.ts').toLowerCase()
                    ],
                    activeFile: undefined
                }
            };

            const files = StateHelper.getFiles(state, rootPath);
            assert.strictEqual(files.length, 1);
            assert.ok(files[0].endsWith('file1.ts'));
        });

        it('should handle empty state', () => {
            const state: State = {
                timestamp: Date.now(),
                source: 'test',
                ide: 'vscode',
                root: '/root',
                editorState: {
                    openedFiles: [],
                    activeFile: undefined
                }
            };
            const files = StateHelper.getFiles(state, '/root');
            assert.strictEqual(files.length, 0);
        });
    });

    describe('isInsideRoot', () => {
        it('should return true if file is inside root', () => {
            const root = '/User/Project';
            const file = '/User/Project/src/file.ts';
            assert.strictEqual(StateHelper.isInsideRoot(root, file), true);
        });

        it('should return false if file is outside root', () => {
            const root = '/User/Project';
            const file = '/User/Other/file.ts';
            assert.strictEqual(StateHelper.isInsideRoot(root, file), false);
        });

        it('should return false for partial directory match', () => {
            const root = '/User/Project';
            const file = '/User/ProjectSuffix/file.ts';
            assert.strictEqual(StateHelper.isInsideRoot(root, file), false);
        });
    });

    describe('checkPathBelongsToState', () => {
        it('should check against state root', () => {
            const rootPath = '/User/Project';
            const state: State = {
                timestamp: Date.now(),
                source: 'test',
                ide: 'vscode',
                root: path.resolve(rootPath).toLowerCase(),
                editorState: {
                    openedFiles: [],
                    activeFile: undefined
                }
            };
            assert.strictEqual(StateHelper.checkPathBelongsToState(state, '/User/Project/file.ts'), true);
            assert.strictEqual(StateHelper.checkPathBelongsToState(state, '/User/Other/file.ts'), false);
        });
    });

    describe('hasIntersection', () => {
        it('should work with two strings', () => {
            assert.strictEqual(StateHelper.hasIntersection('/User/Project', '/User/Project/Sub'), true);
            assert.strictEqual(StateHelper.hasIntersection('/User/Project/Sub', '/User/Project'), true);
            assert.strictEqual(StateHelper.hasIntersection('/User/Project', '/User/Other'), false);
        });

        it('should work with state object', () => {
            const rootPath = '/User/Project';
            const state: State = {
                timestamp: Date.now(),
                source: 'test',
                ide: 'vscode',
                root: path.resolve(rootPath).toLowerCase(),
                editorState: {
                    openedFiles: [],
                    activeFile: undefined
                }
            };
            assert.strictEqual(StateHelper.hasIntersection(state, '/User/Project/Sub'), true);
            assert.strictEqual(StateHelper.hasIntersection(state, '/User/Other'), false);
        });
    });
});

File: jetbrains-plugin/src/test/kotlin/com/oneide/utils/StateHelperTest.kt
package com.oneide.utils

import com.intellij.testFramework.fixtures.BasePlatformTestCase
import com.oneide.models.ActiveFile
import com.oneide.models.EditorState
import com.oneide.models.IdeMetaData
import com.oneide.models.State
import org.junit.Test
import java.nio.file.Paths

/**
 * Unit tests for StateHelper.
 */
class StateHelperTest : BasePlatformTestCase() {

    private lateinit var metaData: IdeMetaData

    override fun setUp() {
        super.setUp()
        metaData = IdeMetaData(project)
    }

    @Test
    fun `test buildState builds a flat state with normalized paths`() {
        val rootPath = "/User/Project"
        val openedFiles = listOf(
            "/User/Project/src/main.kt",
            "/User/Project/README.md"
        )
        val activeFile = ActiveFile(filePath = "/User/Project/src/main.kt", cursor = 10, column = 5)

        val state = StateHelper.buildState(metaData, rootPath, openedFiles, activeFile)

        val rootPathLower = Paths.get(rootPath).toAbsolutePath().normalize().toString().lowercase()
        assertEquals(rootPathLower, state.root)
        assertEquals(2, state.editorState.openedFiles.size)

        val file1 = state.editorState.openedFiles.find { it.endsWith("main.kt") }
        assertNotNull(file1)
        val expectedPath1 = Paths.get("/User/Project/src/main.kt").toAbsolutePath().normalize().toString().lowercase()
        assertEquals(expectedPath1, file1)

        val file2 = state.editorState.openedFiles.find { it.endsWith("readme.md") }
        assertNotNull(file2)
        val expectedPath2 = Paths.get("/User/Project/README.md").toAbsolutePath().normalize().toString().lowercase()
        assertEquals(expectedPath2, file2)
        
        val active = state.editorState.activeFile
        assertNotNull(active)
        assertEquals(expectedPath1, active!!.filePath)
        assertEquals(10, active.cursor)
    }

    @Test
    fun `test buildState filters out files outside of root`() {
        val rootPath = "/User/Project"
        val openedFiles = listOf(
            "/User/Project/in.kt",
            "/User/Other/out.kt"
        )

        val state = StateHelper.buildState(metaData, rootPath, openedFiles, null)
        assertEquals(1, state.editorState.openedFiles.size)
        assertTrue(state.editorState.openedFiles[0].contains("in.kt"))
    }

    @Test
    fun `test getFiles returns files belonging to root`() {
        val rootPath = "/User/Project"
        val rootPathLower = Paths.get(rootPath).toAbsolutePath().normalize().toString().lowercase()
        
        val state = State(
            timestamp = System.currentTimeMillis(),
            source = "test",
            ide = "jetbrains",
            root = rootPathLower,
            editorState = EditorState(
                openedFiles = mutableListOf(
                    Paths.get("/User/Project/file1.kt").toAbsolutePath().normalize().toString().lowercase(),
                    Paths.get("/User/Other/file2.kt").toAbsolutePath().normalize().toString().lowercase()
                ),
                activeFile = ActiveFile(
                     filePath = Paths.get("/User/Project/file1.kt").toAbsolutePath().normalize().toString().lowercase(),
                     cursor = 0
                )
            )
        )

        val files = StateHelper.getFiles(state, rootPath)
        assertEquals(1, files.size)
        assertTrue(files[0].endsWith("file1.kt"))
    }

    @Test
    fun `test getFiles handles empty state`() {
        val state = State(
            timestamp = System.currentTimeMillis(),
            source = "test",
            ide = "jetbrains",
            root = "/root",
            editorState = EditorState(
                openedFiles = mutableListOf()
            )
        )
        val files = StateHelper.getFiles(state, "/root")
        assertEquals(0, files.size)
    }

    @Test
    fun `test isInsideRoot`() {
        val root = "/User/Project"
        val fileInside = "/User/Project/src/file.kt"
        val fileOutside = "/User/Other/file.kt"
        val filePartial = "/User/ProjectSuffix/file.kt"

        assertTrue(StateHelper.isInsideRoot(root, fileInside))
        assertFalse(StateHelper.isInsideRoot(root, fileOutside))
        assertFalse(StateHelper.isInsideRoot(root, filePartial))
    }

    @Test
    fun `test checkPathBelongsToState`() {
        val rootPath = "/User/Project"
        val rootPathLower = Paths.get(rootPath).toAbsolutePath().normalize().toString().lowercase()
        val state = State(
            timestamp = System.currentTimeMillis(),
            source = "test",
            ide = "jetbrains",
            root = rootPathLower,
            editorState = EditorState(
                openedFiles = mutableListOf()
            )
        )

        assertTrue(StateHelper.checkPathBelongsToState(state, "/User/Project/src/file.kt"))
        assertFalse(StateHelper.checkPathBelongsToState(state, "/User/Other/file.kt"))
    }

    @Test
    fun `test hasIntersection`() {
        assertTrue(StateHelper.hasIntersection("/User/Project", "/User/Project/Sub"))
        assertTrue(StateHelper.hasIntersection("/User/Project/Sub", "/User/Project"))
        assertFalse(StateHelper.hasIntersection("/User/Project", "/User/Other"))

        val rootPath = "/User/Project"
        val rootPathLower = Paths.get(rootPath).toAbsolutePath().normalize().toString().lowercase()
        val state = State(
            timestamp = System.currentTimeMillis(),
            source = "test",
            ide = "jetbrains",
            root = rootPathLower,
            editorState = EditorState(
                openedFiles = mutableListOf()
            )
        )

        assertTrue(StateHelper.hasIntersection(state, "/User/Project/Sub"))
        assertFalse(StateHelper.hasIntersection(state, "/User/Other"))
    }
}
